### PR TITLE
KEYCLOAK-14811 KEYCLOAK-14812 RoleProvider, RoleStorageManager

### PR DESCRIPTION
**Pull request by vramik**

### CODE REVIEW

File: services/src/main/java/org/keycloak/storage/RoleStorageManager.java [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+        Stream<RoleModel> local = session.roleLocalStorage().searchForRolesStream(realm, search, first, max);
+        Stream<RoleModel> ext = getEnabledStorageProviders(session, realm, RoleLookupProvider.class).stream()
+                .flatMap(ServicesUtils.timeBound(session,
+                        roleStorageProviderTimeout,
+                        p -> ((RoleLookupProvider) p).searchForRolesStream(realm, search, first, max)));
+
+        return Stream.concat(local, ext);
+    }
+
+    @Override
```
Comment (mhajas) [inline services/src/main/java/org/keycloak/storage/RoleStorageManager.java:179]: This method doesn't exist in RoleProvider interface.

File: services/src/main/java/org/keycloak/storage/RoleStorageManager.java [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    public boolean removeRole(RealmModel realm, RoleModel role) {
+        if (!StorageId.isLocalStorage(role.getId())) {
+            throw new RuntimeException("Federated roles do not support this operation");
+        }
+        return session.roleLocalStorage().removeRole(realm, role);
```
Comment (mhajas) [inline services/src/main/java/org/keycloak/storage/RoleStorageManager.java:184]: We should use `removeRole(role)`.

File: services/src/main/java/org/keycloak/storage/RoleStorageManager.java [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.RoleModel;
+import org.keycloak.models.RoleProvider;
+import org.keycloak.storage.role.RoleLookupProvider;
+import org.keycloak.storage.role.RoleStorageProvider;
+import org.keycloak.storage.role.RoleStorageProviderFactory;
+import org.keycloak.storage.role.RoleStorageProviderModel;
+import org.keycloak.utils.ServicesUtils;
+
+public class RoleStorageManager implements RoleProvider {
```
Comment (mhajas) [inline services/src/main/java/org/keycloak/storage/RoleStorageManager.java:37]: This class doesn't implement methods `removeRoles(realm/client)`.

**Review by mhajas — CHANGES_REQUESTED** [review]
Thanks for the PR @vramik. I tried to update `RealmProvider` so that it doesn't contain any `RoleProvider` methods (just to test if we completely avoid RealmProvider class when working with Roles). There were to places with the usage of deprecated methods that are no longer present in RoleProvider, we should replace them with the corresponding `RoleProvider` method. See following commit: https://github.com/mhajas/keycloak/commit/2e06cc5c2d48af66e4420de901870f567a20c530

There are also some compilation errors in RoleStorageManager, see comments in code.


Otherwise, it looks good, I will do some more testing, but we should have those^ fixed asap so that we can see results from Travis.

**Comment by vramik** [comment]
> I tried to update `RealmProvider` so that it doesn't contain any `RoleProvider` methods (just to test if we completely avoid RealmProvider class when working with Roles). There were to places with the usage of deprecated methods that are no longer present in RoleProvider, we should replace them with the corresponding `RoleProvider` method. See following commit: [mhajas@2e06cc5](https://github.com/mhajas/keycloak/commit/2e06cc5c2d48af66e4420de901870f567a20c530)

Thanks for that, I should have tried it myself, will fix this.

**Review by mhajas — DISMISSED** [review]
LGTM! Thanks, @vramik!

**Comment by vramik** [comment]
@mposolda @pedroigor would you like to review this PR?

**Review by pedroigor — DISMISSED** [review]
@vramik The new tests are failing for Quarkus (the others are expected). Not a blocker for merging, but I appreciate a fix for them.

**Review by martin-kanis — DISMISSED** [review]
@vramik It looks good to me. Thanks!

**Review by sguilhen — CHANGES_REQUESTED** [review]
LGTM, thanks @vramik

**Comment by vramik** [comment]
The test failures on Quarkus are caused by the fact that `DefaultKeycloakSessionFactory.init()` in not called on quarkus. We discussed it with @pedroigor and agreed that we should move the resolving of the default value to the get methods.

**Comment by pedroigor** [comment]
@vmuzikar This type of confusion was expected as we decided to replicate some code from main dist to Quarkus to not mess with the main dist. We should keep an eye on the changes to some classes and avoid more duplications. And fix that in the future so that we have more hooks to customize behavior depending on the stack we're running.

**Review by martin-kanis — APPROVED** [review]

**Comment by vmuzikar** [comment]
@pedroigor Did you mean @vramik? :)

**Review by sguilhen — APPROVED** [review]

**Review by pedroigor — APPROVED** [review]

**Pull request merged**